dnprogs/dncopy/dncopy.1
.TH DNCOPY 1 "January 26 2005" "DECnet utilities"

.SH NAME
dncopy \- Copy files to/from a VMS system

.SH SYNOPSIS
.B dncopy
[options] source dest
.br
.B dncopy
[options] source... directory
.br
.B dntype
source...
.br
Options:
.br
[\-vdisklEVh] [\-m mode] [\-a record attributes] [\-r record format]
[\-b block size] [\-p VMS protection]
.SH DESCRIPTION
.PP
dncopy copies files to and from VMS systems.
.br

.br
Files on VMS systems should be specified in the usual transparent DECnet format
of node"username password"::[directory]file. To protect quotes and dollar signs
from shell expansion I recommend that all VMS file specifications be enclosed
in single quotes (see
.B EXAMPLES
below). If you don't want to type the password on the command-line then
put a hyphen ("-") in its place and you will be prompted for it.
.br

.br
.B dncopy
can copy single files and multiple files. If multiple files are copied the
destination must be a directory but it may be on the local Linux system
or a VMS system. The files to be copied can be a mixture of VMS files and
local files (yes, you can copy from VMS to VMS with this program, though
quite why you would want to I'm not sure)
.br

.br
Wildcards are supported for local and VMS files (of course local wildcards are
expanded by the shell). Remember to use VMS wildcards (*%) on VMS filesystems
and Unix wildcards (*? etc) on Unix files.
.br

.br
The pseudo-filename '-' may be used to represent standard input or standard
output to enable dncopy to be used in a pipeline.
Filenames are changed to lower case when copied from VMS to Linux.
.br

.br
The environment DNCOPY_OPTIONS may be used to provide a default set of
options for copying files. If (for example) you wanted to normally send
files as blocks rather than records you could set DNCOPY_OPTIONS="\-mblock".
Then, to send a file as records you would need to type
.B dncopy \-mrecord myfile.txt vmsbox::
instead.
.br
Options in DNCOPY_OPTIONS may be overridden by options typed on the
command-line except where there is no negating option available (see
.B \-k \-d \-i
). If you put these options in DNCOPY_OPTIONS then to remove them you will
have to override the whole environment variable eg:
.br
$ DNCOPY_OPTIONS="" dncopy myfile.txt vmsbox::
.br

.B dntype
is simply a version of dncopy where the output file is forced to "-". Thus
it takes all the same options as dncopy. It is merely a convenience.

.SH OPTIONS
.TP
.I "\-v"
Verbose operation. The more \-v options are present the more verbose dncopy
will become. One \-v is roughly equivalent to the /LOG qualifier on the DCL
copy command. More than one is really just useful for debugging.
.TP
.I "\-i"
Interactive operation. Prompts before copying a file. This option is roughly
equivalent to the /CONFIRM qualifier on the DCL copy command.
.TP
.I "\-l"
Ignore interlocks on files copied from VMS. This will do its best to read the
data regardless of record or file locking, but it won't always succeed.
.TP
.I "\-s"
Show transfer statistics. This shows the throughput of all copies (in the case
of wildcard transfers) undertaken in K bytes/second. This time does not include
that to establish the connection. eg when sending to VMS the overhead of
creating a NETSERVER process is not included.
.TP
.I "\-k"
Keep version numbers on files copied from VMS systems. By default dncopy will
strip the version number from files because they have no meaning in Linux.
if you specify \-k on the command line then the VMS file will be created with
exactly the same name as it had on the VMS system. To access these files under
Linux you will need to enclose them in quotes. eg less "vmsfile.txt;1".
.TP
.I "\-m {record|block}"
Sets the transfer mode to block or record.
.B record
is the default. Normally
.B record
is what you want but
.B block
is used for sending binary files to the VMS system. Note that if you
use \-mblock to pull files from VMS you may not be able to make sense of
the file on Linux because dncopy will also pull all the VMS internal
structuring of the file as well as the data. Unless you really know what
you are doing \-mblock is only really useful for sending files.
.TP
.I "\-a {none|ftn|cr|prn}"
Sets the carriage control attributes for files copied to a VMS system. The
default is
.B cr.
.TP
.I "\-r {fix|var|vfc|stm}"
Sets the record format for files copied to a VMS system. The default is
.B stm.
In fact the default is STREAMLF as this corresponds to the format of files
on Unix systems and so is the least likely to result in file corruption.
If you are sending a pure text file then
.B var
or
.B vfc
may be more appropriate for your application.
.B fix
may be useful for block-structures files and data files.
.TP
.I \-b N
Set the block size for transfers. The default is high enough for all record
structured files. if you are sending a file with
.B \-mblock
then you should use this to set the block size of the file to be created at
the VMS end. When sending block files the default changes to 512 as this is
typically what you would want (I hope).
.TP
.I "\-d"
Remove any trailing CR characters at the end of a line. This is useful for
sending DOS files to VMS. Only works when transferring in record mode.
.TP
.I "\-p 'protection'"
When sending files to VMS, sets the protection of the newly created remote file.
Without this option, VMS will set the protection to be the default for the remote user.  The protection should be in VMS-style format and enclosed in single 
quotes to protect it from the shell (see example). 
.br
This option is ignored when copying from VMS.
.TP
.I "\-P"
Queue the file for printing to SYS$PRINT when it arrives at the VMS end.
.TP
.I "\-D"
Delete the file when it is closed. This is only really useful in conjunction
with 
.I \-P.
.TP
.I "\-T connect timeout"
Specifies the maximum amount of time the command will wait to establish a connection
with the remote node. a 0 here will cause it to wait forever. The default is 60 seconds
.TP
.I \-E
Ignore errors opening output files. This is handy if you are sending a lot
of Unix files to VMS, some of which have illegal filenames (eg ~ backup files).
dncopy will report an error for each file but continue sending.
.TP
.I \-h \-?
Displays help for using the command.
.TP
.I \-V
Show the version of the tools package that dncopy comes from.
.br
.SH ENVIRONMENT VARIABLE
You can put your most commonly used defaults in the environment variable
.B DNCOPY_OPTIONS
eg:
.br

.br
bash or ksh:
.br
$ DNCOPY_OPTIONS="\-mblock \-anone \-b1024" ; export DNCOPY_OPTIONS
.br

.br
csh or tcsh:
.br
$ setenv DNCOPY_OPTIONS "\-mblock \-anone \-b1024"
.br

.br
makes dncopy send files as 1024 byte blocks with no carriage control. You can
override these options by specifying replacements on the command-line as usual.
Be aware that some options have no converse (eg \-i \-k \-d \-v) so if you put
these in DNCOPY_OPTIONS you cannot disable them without deassigning the
variable.
.br

.SH EXAMPLES
.br
.br
Copy LOGIN.COM from the VMS system "tramp" to Linux as mylogin.com
.nf
.br
.PP
    dncopy 'tramp"christine pjc123"::login.com' mylogin.com
.br

.br
Copy all .TXT files from the VMS directory SYS$SYSDEVICE:[WP] to /tmp:
.br
.PP
    dncopy 'tramp"christine pjc123"::sys$sysdevice:[wp]*.txt' /tmp
.br

.br
Copy an executable to VMS:
.br
.PP
    dncopy \-mblock test.exe 'tramp"christine pjc123"::[.BIN]'
.br

.PP
.br
Copy a file to VMS and set its protection
.br
.PP
    dncopy secret.dat  marsha:: \-p (s:, o:rwed, g:re, w:)'
.br
.br
.PP
.br
Display the contents of LOGIN.COM:
.br
.PP
    dtype 'trisha"christine \-"::login.com'
.br
    You will then be prompted for a password


.SH HELPFUL HINTS
For fetching files the defaults should serve for most purposes. Most VMS files
are record orientated and
.B \-mrecord
is the default transfer mode. It is rare you will need to fetch files using
.B \-mblock
because you will get all the record control information downloaded too and
that probably isn't any use to you.
.br

.br
Sending files is more complex because VMS supports far more attributes
than Linux so you will need to know something about the file you are
sending. Text files should be OK with the defaults unless you need to change
the format from the default
.B STREAMLF
to
.B VFC
or
.B VAR.
Carriage control can also be specified if you want to be that picky.

Binary files may often need to be sent \-mblock to be useful at the VMS end,
You will probably want to specify a block size with the
.B \-b
option. The default is 512 which is fairly useful but if you are sending
(say) a saveset 8192 or 32256 may be required. Trial-and-error may
be the only way in some cases unless you know the file contents very well.
If you really don't know what to do, just send it
.B \-mblock
and use the set file/attr command to massage it on the VMS end until you are
happy with it. (If you are using VMS earlier than 6.1 then you will need
the freeware FILE utility to do this)
.SH SEE ALSO
.BR dntype "(1), " dndir "(1), " dndel "(1), " dntask "(1), " dnsubmit "(1), " dnprint "(1)"


dnprogs/NEWS
dnprogs (2.61) unstable; urgency=low

  * Include libvaxdata/src/test.c in the distribution tarball
  * Fix password length in DAP connections. It was incorrectly set at 12 
    rather than 40
 
 -- Chrissie Caulfield <chrissie@debian.org>  Sun, 31 Mar 2013 11:18:32 +0000

dnprogs (2.60+x32) unreleased; urgency=low

  * Switch from using deprecated sysctl(2) to writing to /proc/sys.  This
    fixes a build failure on x32.

 -- Daniel Schepler <schepler@debian.org>  Tue, 12 Feb 2013 12:21:31 -0800

dnprogs (2.60) unstable; urgency=low

  * Fixed lots of Debian things that caused lintian problems and 
    upgraded policy version
    Closes: #673408
  * Fix some warnings that might be bugs
  * Incorporate NMU into main tree.

 -- Chrissie Caulfield <chrissie@debian.org>  Sat, 26 May 2012 09:54:41 +0100

dnprogs (2.59+nmu2) unstable; urgency=medium

  * NMU
  * dnprogs binaries are Linux-only, so state that. Move dnet-common back
    to binary-all. Closes: #665886

 -- Steve McIntyre <93sam@debian.org>  Mon, 21 May 2012 17:43:14 +0100

dnprogs (2.59+nmu1) unstable; urgency=low

  * NMU
  * Reduce libdnet's recommends on dnet-common to suggests again.
    Closes: #655740. Adapted patch from Jonathan Nieder.

 -- Steve McIntyre <93sam@debian.org>  Fri, 18 May 2012 13:29:53 +0100

dnprogs (2.59) unstable; urgency=low

  * Changed package dnet-common from arch all to any and
    updated dependecies. (Closes: #665886)

 -- Chrissie Caulfield <chrissie@debian.org>  Fri, 27 Apr 2012 09:30:19 +0100

dnprogs (2.58) unstable; urgency=medium

  * Fixed lintian warnings:
    - Fixed changelog entry of last release to have correct line length.
    - Corrected debconf dependency in dnet-common.
    - Corrected versioned dependency in dnprogs on dnet-common.
    - Updated existing overrides to match new syntax.
  * Updated linkage of libdnet_daemon and libdap. (Closes: #646353)
  * Fixed compile warning in dndir when using -Werror=format-security
    (Closes: #646613)
  * Fixed typo in debian/dnet-common.config
  * Do not try to install files into packages on FreeBSD which are
    not build on FreeBSD.
  * Link dapfs with libpthread for new libfuse compatibility
  * Updated nl.po (Closes: #652349)
  * Added pl.po (Closes: #661894)
  * Added it.po (Closes: #663135)
  * Change module-init-tools dependency to kmod (Closes: #662684)

 -- Chrissie Caulfield <chrissie@debian.org>  Sat, 10 Mar 2012 10:32:08 +0000

dnprogs (2.57) unstable; urgency=medium

  [ Chrissie Caulfield ]
  * Completed removal of modutils configuration file code
  * Fix presence of CVS directories in tarball
  * updated 'clean' target

  [ Philipp Schafft ]
  * Use /dev/pts PTYs on 64 bit systems (fix makefile bug)
  * Updated translations Debconf templates:
    - German translation (Closes: #615268)
    - French translation (Closes: #615039)
    - Danish translation (Closes: #633984) (added)
    - Slovak translation (Closes: #639629) (added)
  * Look in yet another place for libcrypt (better fix than from NMUs)
  * Added a comment about the detecting of libcrypt. (See: #629664)
  * "ethernet" was capitalized in extended description (Closes: #636798)
  * Fixed some lintian warnings (mainly new rules targets).
  * dnet-common: Do not configure DECnet unless the user requested it.
    (Removed the bad workaround done by NMU) (See: #637179)
  * dnet-common: Added preinst script to clean up the broken config files
    from the last NMU. (Closes: #640861)
  * Upgraded libdnet's suggest to dnet-common to recommend again
    (Closes: #608807)

 -- Chrissie Caulfield <chrissie@debian.org>  Thu,  8 Sep 2011 16:05:47 +0200

dnprogs (2.56.1+nmu1) unstable; urgency=low

  * Non-maintainer upload.
  * libdnet: Only suggest dnet-common package (Closes: #608807)
  * dnet-common: Do not change MAC addresses unless user chooses to
    configure DECnet (Closes: #637179)
  * dnet-common: Restore ability to configure DECnet name and address
    through debconf
  * dnet-common: Remove obsolete modutils configuration file

 -- Ben Hutchings <ben@decadent.org.uk>  Thu, 25 Aug 2011 14:26:12 +0100

dnprogs (2.56.1) unstable; urgency=low

  * Non-maintainer upload.
  * Add patch from Ubuntu to fix FTBFS with libcrypt moved to the
    multiarch path.

 -- Aurelien Jarno <aurel32@debian.org>  Tue, 09 Aug 2011 11:58:06 +0200


dnprogs (2.56) unstable; urgency=low

  * Search for libgcrypt in 64bit library paths, too. (thanks to Kamil)
  * Package dnet-common now asks if network interfaces should be configrued
    for DECnet (Closes: #608807)
    Updated translations Debconf templates:
    - Russian translation (Closes: #610642)
    - Portuguese translation (Closes: #611034)
    - Swedish translation (Closes: #611296)
    - Czech translation (Closes: #611430)
    - Japanese translation (Closes: #611448)
    - Spanish translation (Closes: #611547)
  * Build against version 25 of FUSE
    Closes: #598831
  * Fixed some lintian warnings (typos and such).
  * Raised priority of package libdnet from extra to optional
    to comply policy manual section 2.5 "Priorities".

 -- Chrissie Caulfield <chrissie@debian.org>  Tue, 22 Feb 2011 10:38:23 +0000

dnprogs (2.55) unstable; urgency=low

  * Build against version 25 of FUSE
    Closes: #598831

 -- Chrissie Caulfield <chrissie@debian.org>  Tue, 5 Oct 2010 09:23:12 +0100

dnprogs (2.54) unstable; urgency=low

  * Don't build libvaxdata test program by default.
    Closes: #598752
 
 -- Chrissie Caulfield <chrissie@debian.org>  Fri, 1 Oct 2010 20:44:09 +0100

dnprogs (2.53) unstable; urgency=low

  * Done some more porting to FreeBSD, corrected Build-Depends for hurd-i386
    Closes: #588185
  * Added new tool node(1) - Node name lookup tool
  * Added support to libdnet_daemon to read /etc/nodes.{allow,deny}
  * Fixed DoS bug in dnet_daemon()
    Closes: #592015 
  * Update libvaxdata to version 1.2, fixes double conversions

 -- Christine Caulfield <Christine.Caulfield@googlemail.com>  Thu, 19 Jul 2010 10:19:32 +0100


dnprogs/dncopy/dncopy.cc
/******************************************************************************
    (c) 1998-2008 Christine Caulfield               christine.caulfield@googlemail.com

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
 ******************************************************************************
 */
#include <sys/types.h>
#include <sys/time.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <netdnet/dn.h>
#include <regex.h>
#include "connection.h"
#include "protocol.h"

#include "file.h"
#include "dnetfile.h"
#include "unixfile.h"

static bool  dntype = false;
static bool  cont_on_error = false;

// Prototypes
static void usage(char *name, int dntype, FILE *f);
static file *getFile(const char *name, int verbosity);
static void get_env_as_args(char **argv[], int &argc, char *env);
static void do_options(int argc, char *argv[],
		       int &rfm, int &rat, int &org,
		       int &interactive, int &keep_version, int &user_bufsize,
		       int &remove_cr, int &show_stats, int &verbose,
		       int &flags, char *protection, int &connect_timeout);

// Start here:
int main(int argc, char *argv[])
{
    file *out;
    file *in;
    int   num_input_files;
    int   filenum;
    char *buf;
    int   rat = file::RAT_DEFAULT;
    int   rfm = file::RFM_DEFAULT;
    int   org = file::MODE_RECORD;
    int   user_bufsize = -1;
    int   bufsize = 16384;
    int   verbose = 0;
    int   interactive = FALSE;
    int   keep_version = FALSE;
    int   last_infile;
    int   remove_cr = 0;
    int   show_stats = 0;
    int   printfile = 0;
    int   flags = 0;
    int   connect_timeout = 60;
    char  opt;
    char  protection[255]={'\0'};
    struct timeval start_tv;
    unsigned long long bytes_copied = 0;

    if (argc < 2)
    {
	usage(argv[0], dntype, stdout);
	return 0;
    }

    // See if we are dntype or dncopy
    if ((strstr(argv[0], "dntype") != NULL))
    {
        dntype = TRUE;
    }

// If there is a DNCOPY_OPTIONS environment variable then parse that first
    char **env_argv;
    int    env_argc;
    get_env_as_args(&env_argv, env_argc, getenv("DNCOPY_OPTIONS"));
    if (env_argc)
	do_options(env_argc, env_argv,
		   rfm, rat,org,
		   interactive, keep_version, user_bufsize,
		   remove_cr, show_stats, verbose, flags, protection, connect_timeout);


// Parse the command-line options
    do_options(argc, argv,
	       rfm, rat,org,
	       interactive, keep_version, user_bufsize,
	       remove_cr, show_stats, verbose, flags, protection, connect_timeout);

    // Work out the buffer size. The default for block transfers is 512
    // bytes unless the user specified otherwise.
    if (user_bufsize > -1)
        bufsize = user_bufsize;
    else
      if (org == file::MODE_BLOCK)
	  bufsize = 512;

    // If the user wants a block transfer and did not specify
    // a record format then default to FIXed length records
    // with no carriage control.
    if (org == file::MODE_BLOCK)
    {
	if (rat == file::RAT_DEFAULT) rat = file::RAT_NONE;
        if (rfm == file::RFM_DEFAULT) rfm = file::RFM_FIX;
    }

    // Get the input file name(s)
    num_input_files = argc - optind - 1;

    // If the command is dntype then output to stdout
    if (dntype)
    {
	out = getFile("-", verbose);
	last_infile = argc;
    }
    else
    {
	if ( (argc - optind) < 2) // Are there enough args?
	{
	    usage(argv[0], dntype, stderr);
	    return 0;
	}
	out = getFile(argv[argc-1], verbose);
	last_infile = argc-1;
    }

    // If there are multiple input files or the one input file is a wildcard
    // then the output must be a directory.
    if (num_input_files > 1 && !dntype)
    {
	if (!out->isdirectory())
	{
	    fprintf(stderr, "Output must be a directory\n");
	    return 3;
	}
    }

    // Allocate transfer buffer. Always allocate the transfer size to allow
    // for the remote end streaming data to us.
    buf = (char *)malloc(65536);
    if (!buf)
    {
	fprintf(stderr, "Cannot allocate transfer buffer");
	out->close();
	return 2;
    }

    // Set up the network links if necessary
    if (out->setup_link(bufsize, rfm, rat, org, flags, connect_timeout))
    {
	out->perror("Error setting up output link");
	return 1;
    }

    // Reduce the buffer size to the biggest the output host can handle.
    bufsize = out->max_buffersize(bufsize);

    for (filenum = optind; filenum < last_infile; filenum++)
    {
	    in = getFile(argv[filenum], verbose);

	// Now we have the first file name, if it is the only input file then
	// we can check to see if it is a wildcard. If that is so then the
	// output must be a directory.
	if (in->iswildcard())
	{
	    if (!out->isdirectory() && !dntype)
	    {
		fprintf(stderr, "Output must be a directory\n");
		return 3;
	    }
	}

	if (in->setup_link(bufsize, rfm, rat, org, flags, connect_timeout))
	{
	    in->perror("Error setting up input link");
	    return 1;
	}

	// Start the timer after the link setup. You may think that's
	// a bit of a cheat, well - tough.
	if (show_stats)
	    gettimeofday(&start_tv, NULL);

	// Copy the file(s)
	do
	{
	    int buflen;
	    int blocks = 0;
	    int do_copy = !interactive;


	    // Open the input file
	    if (in->open("r"))
	    {
		in->perror("Error opening file for input");
		return 1;
	    }

	    if (interactive)
	    {
		char response[80];

		if (dntype)
		    printf("Type %s ? ",
		       in->get_printname());
		else
		    printf("Copy %s to %s ? ",
			   in->get_printname(),
			   out->get_printname(in->get_basename(keep_version)));

		fgets(response, sizeof(response), stdin);
		if (tolower(response[0]) == 'y')
		    do_copy = TRUE;
	    }

	    if (do_copy)
	    {
		// If the output is a directory so we need to add the
		// input file's basename to it.
		out->set_protection(protection);
		if (out->isdirectory())
		{
		    if (out->open(in->get_basename(keep_version), "w+"))
		    {
			out->perror("Error opening file for output");
			in->close();
			if (cont_on_error)
			    continue;
			else
			    return 1;
		    }
		}
		else
		{
		    if (out->open("w+"))
		    {
			out->perror("Error opening file for output");
			in->close();
			return 1;
		    }
		}

		if (dntype && verbose) printf("\n%s\n\n", in->get_printname());

		// Copy the data
		while ( ((buflen = in->read(buf, bufsize))) >= 0 )
		{
		    // Remove trailing CRs if required
		    if (remove_cr &&
			org == file::MODE_RECORD &&
			buf[buflen-2] == '\r')
		    {
			// CR is before the LF in the buffer.
			buf[buflen-2] = buf[buflen-1];
			buflen--;
		    }

		    if (out->write(buf, buflen) < 0)
		    {
			out->perror("Error writing");
			in->close();
			return 3;
		    }
		    blocks++;
		    bytes_copied += buflen;
		}

		// If we finished with an error then display it
		if (!in->eof())
		{
		    in->perror("Error reading");
		    out->close();
		    return 3;
		}

		// Set the file protection.
		if (out->set_umask(in->get_umask()) && !dntype)
		{
		    out->perror("Error setting protection");
		    // Non-fatal error this one.
		}
		if (!dntype)
		{
		    if (out->close())
		    {
			out->perror("Error closing output");
			return 3;
		    }
		}

		// Log the operation if we were asked
		if (verbose && !dntype)
		    printf("'%s' copied to '%s', %d %s\n",
			   in->get_printname(),
			   out->get_printname(),
			   blocks,
			   in->get_format_name());

	    }
	    if (in->close())
	    {
		out->perror("Error closing input");
		return 3;
	    }
	}
	while(in->next());
    }

    // Show stats
    if (show_stats)
    {
	struct timeval stop_tv;
	long centi_seconds;
	double rate;
	double show_secs;

	gettimeofday(&stop_tv, NULL);
	centi_seconds = (stop_tv.tv_sec - start_tv.tv_sec) * 100 +
	    (stop_tv.tv_usec - start_tv.tv_usec) / 10000;
	show_secs = (double)centi_seconds/100.0;

	rate = (double)(bytes_copied/1024) / (double)centi_seconds * 100.0;
	printf("Sent %lld bytes in %1.2f seconds: %4.2fK/s\n",
	       bytes_copied, show_secs, rate);
    }
}

// Print a usage message. We can be called as dncopy or dntype so adapt
// accordingly
static void usage(char *name, int dntype, FILE *f)
{
    fprintf(f, "\nusage: %s [OPTIONS] infile", name);
    if (!dntype)
    {
	fprintf(f, " outfile");
	fprintf(f, "\n   or: %s [OPTIONS] infiles...", name);
	fprintf(f, " directory");
    }

    fprintf(f, "\n\n");
    fprintf(f, " Options\n");
    fprintf(f, "  -? -h        display this help message\n");
    fprintf(f, "  -v           verbose operation.\n");
    fprintf(f, "  -s           show transfer statistics.\n");
    if (!dntype)
    {
        fprintf(f, "  -i           interactive. Prompt before copying\n");
        fprintf(f, "  -k        (r)keep version numbers on files\n");
        fprintf(f, "  -m <mode>    access mode: (record, block)\n");
        fprintf(f, "  -a <att>  (s)record attributes (none, ftn, cr, prn)\n");
        fprintf(f, "  -r <fmt>  (s)record format (fix, var, vfc, stm)\n");
        fprintf(f, "  -b <n>       use a block size of <n> bytes\n");
        fprintf(f, "  -d        (s)remove trailing CR on record (DOS file transfer)\n");
        fprintf(f, "  -l        (r)ignore interlocks on remote file\n");
	fprintf(f, "  -P        (s)print file to SYS$PRINT\n");
	fprintf(f, "  -D        (s)delete file on close. Only really useful with -P\n");
	fprintf(f, "  -T <secs>    connect timeout in seconds (default 60)\n");
        fprintf(f, "  -V           show version number\n");
        fprintf(f, "\n");
        fprintf(f, " (s) - only useful when sending files to VMS\n");
        fprintf(f, " (r) - only useful when receiving files from VMS\n");
    }
    else
    {
        fprintf(f, "  -i           interactive. Prompt before displaying\n");
    }
    fprintf(f, "\n");
    fprintf(f, "NOTE: It is a good idea to put VMS filenames in single quotes\n");
    fprintf(f, "to stop the shell from swallowing special characters such as double\n");
    fprintf(f, "quotes and dollar signs. eg:\n");
    fprintf(f, "\n");
    fprintf(f, "%s 'mynode\"christine password\"::sys$manager:sylogin.com'", name);
    if (!dntype)
    {
	fprintf(f, " .");
    }
    fprintf(f, "\n\n");
}

// Run through the file types and return an object that matches the
// type of the name we were passed.
static file *getFile(const char *name, int verbosity)
{
    if (dnetfile::isMine(name))
	return new dnetfile(name, verbosity);

    // Lots of opportunities for other file types here...


    // Default to a local file.
    return new unixfile(name);
}

static void get_env_as_args(char **argv[], int &argc, char *env)
{
    int count = 0;
    char *ptr;

    argc = 0;
    if (!env) return; // No variable.

    // Take a copy of the arglist as we mangle it.
    char *arglist = (char *)malloc(strlen(env)+1);
    strcpy(arglist, env);

// Quickly run through the variable to see how many options there are.
    ptr = strtok(arglist, " ");
    while(ptr)
    {
	count++;
	ptr = strtok(NULL, " ");
    }

    *argv = (char **)malloc((sizeof(char *) * (count+2)));
    argc = count+1;
    char **pargv = *argv;
    strcpy(arglist, env);

// Now build the array of args, starting at 1
// 'cos 0 is the program name, remember?
    count = 1;
    ptr = strtok(arglist, " ");
    while(ptr)
    {
	pargv[count] = (char *)malloc(strlen(ptr)+1);
	strcpy(pargv[count], ptr);
	ptr = strtok(NULL, " ");
	count++;
    }
    pargv[count] = NULL; // Make sure the last element is NULL
    free(arglist);
}

// Process the options
static void do_options(int argc, char *argv[],
		       int &rfm, int &rat, int &org,
		       int &interactive, int &keep_version, int &user_bufsize,
		       int &remove_cr, int &show_stats, int &verbose,
		       int &flags, char *protection, int &connect_timeout)
{
    int opt;
    opterr = 0;
    optind = 0;
    while ((opt=getopt(argc,argv,"?Vvhdr:a:b:kislm:p:PDET:")) != EOF)
    {
	switch(opt) {
	case 'h':
	    usage(argv[0], dntype, stdout);
	    exit(0);

	case '?': // Called if getopt doesn't recognise the option
	    usage(argv[0], dntype, stderr);
	    exit(0);

	case 'v':
	    verbose++;
	    break;

	case 'l':
	    flags |= file::FILE_FLAGS_RRL;
	    break;

	case 'E':
	    cont_on_error = true;
	    break;

	case 'T':
	    connect_timeout = atoi(optarg);
	    break;

	case 'r':
	    if (tolower(optarg[0]) == 'f') rfm = file::RFM_FIX;
	    if (tolower(optarg[0]) == 's') rfm = file::RFM_STM;
	    if (!strncmp(optarg, "va", 2)) rfm = file::RFM_VAR;
	    if (!strncmp(optarg, "vf", 2)) rfm = file::RFM_VFC;
	    if (rfm == file::RFM_DEFAULT)
	    {
		fprintf(stderr, "Invalid record format string\n");
		fprintf(stderr, "%s -h for more information\n", argv[0]);
		exit(1);
	    }
	    break;

	case 'a':
	    if (tolower(optarg[0]) == 'f') rat = file::RAT_FTN;
	    if (tolower(optarg[0]) == 'c') rat = file::RAT_CR;
	    if (tolower(optarg[0]) == 'p') rat = file::RAT_PRN;
	    if (tolower(optarg[0]) == 'n') rat = file::RAT_NONE;
	    if (rat == file::RAT_DEFAULT)
	    {
		fprintf(stderr, "Invalid record attributes string\n");
		fprintf(stderr, "%s -h for more information\n", argv[0]);
		exit(1);
	    }
	    break;

	case 'm':
	    org = file::MODE_DEFAULT; // Set to default so w can catch errors
	    if (tolower(optarg[0]) == 'r') org = file::MODE_RECORD;
	    if (tolower(optarg[0]) == 'b') org = file::MODE_BLOCK;
	    if (org == file::MODE_DEFAULT)
	    {
		fprintf(stderr, "Invalid transfer mode\n");
		fprintf(stderr, "%s -h for more information\n", argv[0]);
		exit(1);
	    }
	    break;

	case 'i':
	    interactive++;
	    break;

	case 's':
	    show_stats++;
	    break;

	case 'k':
	    keep_version = TRUE;
	    break;

	case 'P':
	    flags |= file::FILE_FLAGS_SPOOL;
	    break;

	case 'D':
	    flags |= file::FILE_FLAGS_DELETE;
	    break;

	case 'b':
	    user_bufsize = atoi(optarg);
	    break;

	case 'p':
	    strcpy(protection, optarg);
	    strcpy(protection, optarg);
	    {
                    // Validate protection
                    dap_protect_message prot;
                    if (prot.set_protection(protection) == -1)
                    {
                            fprintf(stderr, "Invalid VMS protection string\n");
                            exit(1);
                    }
            }
	    break;

	case 'd':
	    remove_cr++;
	    break;

	case 'V':
	    printf("\ndncopy/dntype from dnprogs version %s\n\n", VERSION);
	    exit(1);
	    break;
	}
    }
}
